src/category_theory/limits/shapes/nat_iso/binary_products.lean

```diff
@@ -0,0 +1,130 @@
+/-
+Copyright (c) 2019 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.limits.shapes.binary_products
+
+/-!
+Constructing binary products from specified objects, and characterisations of the morphisms
+out them.
+-/
+
+universes v u
+
+open category_theory
+open opposite
+
+namespace category_theory.limits
+
+variables {C : Type u} [𝒞 : category.{v} C]
+include 𝒞
+
+open walking_pair
+
+local attribute [tidy] tactic.case_bash
+
+/--
+We characterise `F.cones` objectwise for a functor `F` on the walking pair.
+-/
+@[simps]
+def walking_pair_cones_equiv {Q : C} (F : discrete walking_pair.{v} ⥤ C) :
+  F.cones.obj (op Q) ≅ ((Q ⟶ F.obj left) : Type v) × ((Q ⟶ F.obj right) : Type v) :=
+{ hom := λ c, (c.app left, c.app right),
+  inv := λ f, { app := λ j, walking_pair.cases_on j f.1 f.2 } }
+
+/--
+`is_binary_product X Y P` asserts that there is an isomorphism of hom-spaces
+`(Q ⟶ P) ≅ (Q ⟶ X) × (Q ⟶ Y)`, natural in `Q`.
+-/
+structure is_binary_product (X Y P : C) :=
```

Comment:
This is certainly the more readable statement. But I wonder if taking the more inherently functorial approach would buy us something down the road. I.e. declare this to be an iso between `yoneda.obj P` and the product of the yoneda functors of `X` and `Y` in the functor category. Of course the naturality part of that iso wouldn't break down so nicely into two parts, but maybe simp-lemma can take care of that.